packages/build/src/plugins/compatibility.ts

```diff
@@ -0,0 +1,106 @@
+ *   - This is the most recent version compatible with this site
+ *   - This is the same logic except it does not use version pinning
+ *   - This is only used to print a warning message when the `compatibleVersion`
+ *     is older than the currently used version.
+ */
+export const getExpectedVersion = async function ({
+  versions,
+  nodeVersion,
+  packageJson,
+  packagePath,
+  buildDir,
+  pinnedVersion,
+}: {
+  versions: PluginVersion[]
+  /** The package.json of the repository */
+  packageJson: PackageJson
+  packagePath?: string
+  buildDir: string
+  nodeVersion: string
+  pinnedVersion?: string
+}) {
+  const { version, conditions = [] } = await getCompatibleEntry({
+    versions,
+    nodeVersion,
+    packageJson,
+    packagePath,
+    buildDir,
+    pinnedVersion,
+  })
+
+  // Retrieve warning message shown when using an older version with `compatibility`
+  const compatWarning = conditions.map(({ type, condition }) => CONDITIONS[type].warning(condition as any)).join(', ')
+  return { version, compatWarning }
+}
+
+/**
+ *  This function finds the right `compatibility` entry to use with the plugin.
+ *   - `compatibitlity` entries are meant for backward compatibility
+ *     Plugins should define each major version in `compatibility`.
+ *   - The entries are sorted from most to least recent version.
+ *   - After their first successful run, plugins are pinned by their major
+ *     version which is passed as `pinnedVersion` to the next builds.
+ *  When the plugin does not have a `pinnedVersion`, we use the most recent
+ *  `compatibility` entry with a successful condition.
+ *  When the plugin has a `pinnedVersion`, we do not use the `compatibility`
+ *  conditions. Instead, we just use the most recent entry with a `version`
+ *  matching `pinnedVersion`.
+ *  When no `compatibility` entry matches, we use:
+ *   - If there is a `pinnedVersion`, use it unless `latestVersion` matches it
+ *   - Otherwise, use `latestVersion`
+ */
+const getCompatibleEntry = async function ({
+  versions,
+  nodeVersion,
+  packageJson,
+  packagePath,
+  buildDir,
+  pinnedVersion,
+}: {
+  versions: PluginVersion[]
+  packageJson: PackageJson
+  buildDir: string
+  nodeVersion: string
+  packagePath?: string
+  pinnedVersion?: string
+}): Promise<Pick<PluginVersion, 'conditions' | 'version'>> {
+  const compatibleEntry = await pLocate(versions, async ({ version, overridePinnedVersion, conditions }) => {
+    // Detect if the overridePinnedVersion intersects with the pinned version in this case we don't care about filtering
+    const overridesPin = Boolean(
+      pinnedVersion && overridePinnedVersion && semver.intersects(overridePinnedVersion, pinnedVersion),
+    )
+
+    // ignore versions that don't satisfy the pinned version here if a pinned version is set
+    if (!overridesPin && pinnedVersion && !semver.satisfies(version, pinnedVersion, { includePrerelease: true })) {
```

Comment:
Does this means people will be prompted to upgrade to prerelease versions?

Reply:
No this part was already there I've just simplified the whole logic here: https://github.com/netlify/build/blob/main/packages/build/src/plugins/compatibility.js#L46

Instead of looping and filtering multiple times it's now just one loop:

https://github.com/netlify/build/blob/main/packages/build/src/plugins/compatibility.js#L44-L69